hadoop-shims/hadoop2.8/src/main/java/org/apache/uniffle/hadoop/shim/HadoopShim28.java

```diff
@@ -0,0 +1,62 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.uniffle.hadoop.shim;
+
+import java.io.IOException;
+import java.lang.reflect.Constructor;
+import org.apache.commons.logging.Log;
+import org.apache.commons.logging.LogFactory;
+import org.apache.hadoop.mapred.JobConf;
+import org.apache.hadoop.mapreduce.TaskAttemptID;
+import org.apache.hadoop.mapreduce.task.reduce.ShuffleClientMetrics;
+import org.apache.hadoop.mapreduce.v2.app.AppContext;
+import org.apache.hadoop.mapreduce.v2.app.client.ClientService;
+import org.apache.hadoop.mapreduce.v2.app.rm.RMContainerAllocator;
+import org.apache.hadoop.yarn.api.protocolrecords.AllocateResponse;
+import org.apache.hadoop.yarn.exceptions.YarnException;
+
+public class HadoopShim28 extends HadoopShim {
+
+  private static final Log LOG = LogFactory.getLog(HadoopShim28.class);
+
+  public ShuffleClientMetrics createShuffleClientMetrics(TaskAttemptID taskAttemptID, JobConf jobConf) {
+    try {
+      Constructor constructor =
+          ShuffleClientMetrics.class.getConstructor(new Class[] {TaskAttemptID.class, JobConf.class});
+      constructor.setAccessible(true);
+      return (ShuffleClientMetrics) constructor.newInstance(taskAttemptID, jobConf);
+    } catch (Exception e) {
+      LOG.warn("Construct ShuffleClientMetrics fail, caused by {}", e);
+      return null;
+    }
+  }
+
+  @Override
+  public RMContainerAllocator createRMContainerAllocator(ClientService clientService, AppContext context) {
```

Comment:
Is it the same as 3.2's implement? Could we put this into `DefaultHadoopShim`?

Reply:
OK, I get your point. The interface is different. We can use reflection to solve this problem. We don't add the add the interface `HadoopShim`.